File: src/EulerLoop.java
import java.util.ArrayList;
import java.util.Stack;

public class EulerLoop {

    private Graph G;

    public EulerLoop(Graph g) {
        this.G = g;
    }

    public boolean hasEulerLoop() {
        CC cc = new CC(G);
        if (cc.getCcount() > 1) return false;

        for (int v = 0; v < G.V(); v++) {
            if (G.degree(v) % 2 == 1) return false;

        }
        return true;
    }

    public ArrayList<Integer> result() {
        ArrayList<Integer> res = new ArrayList<>();
        if (!hasEulerLoop()) return res;
        Graph g = (Graph) G.clone();
        Stack<Integer> stack = new Stack<>();
        int curv = 0;
        stack.push(curv);
        while (!stack.empty()) {
            if (g.degree(curv) != 0) {
                stack.push(curv);
                int w = g.adj(curv).iterator().next();
                g.removeEdge(curv, w);
                curv = w;
            } else {
                res.add(curv);
                curv = stack.pop();
            }
        }
        return res;
    }
    public static void main(String[] args){

        Graph g = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/g.txt");
        EulerLoop el = new EulerLoop(g);
        System.out.println(el.result());

        Graph g2 = new Graph("/Users/hui/Desktop/java/play-with-graph-algorithme/g4.txt");
        EulerLoop el2 = new EulerLoop(g2);
        System.out.println(el2.result());
    }
}

File: src/Chapter09EulerLoop/python/EulerLoop.py
from Chapter02GraphDFS.python.adjSet import adjSet as Graph
from Chapter04GraphBFS.python.CC import CC
from copy import copy


class EulerLoop:
    def __init__(self, G):
        self._G = G

    def has_euler_loop(self):
        cc = CC(self._G)
        if cc.ccount > 1:
            return False

        for v in range(self._G.V()):
            if self._G.degree(v) % 2 != 0:
                return False

        return True

    def result(self):
        res = []
        if not self.has_euler_loop:
            return res
        stack = []
        currv = 0
        stack.append(1)
        newg = copy(self._G)
        while len(stack) > 0:
            if newg.degree(currv) > 0:
                stack.append(currv)
                w = list(newg.adj(currv))[0]
                newg.remove_edge(currv, w)
                currv = w
            else:
                res.append(currv)
                currv = stack.pop()
        return res



if __name__ == '__main__':
    filename = '../g.txt'
    g = Graph(filename)
    eluer_loop = EulerLoop(g)
    print(eluer_loop.result())

    filename = '../g2.txt'
    g = Graph(filename)
    eluer_loop = EulerLoop(g)
    print(eluer_loop.result())